fastfood/food.py

```diff
@@ -122,17 +79,103 @@ def process_stencil(cookbook, cookbook_name, templatepack,
             if force:
                 LOG.warning("Forcing overwrite of existing file %s.",
                             target_path)
+            elif target_path in written_files:
+                LOG.warning("Previous stencil has already written file %s.",
+                            target_path)
             else:
                 print("Skipping existing file %s" % target_path)
                 LOG.info("Skipping existing file %s", target_path)
                 continue
 
-        with open(target_path, 'w') as newfile:
+        with open(target_path, open_mode) as newfile:
             print("Writing rendered file %s" % target_path)
             LOG.info("Writing rendered file %s", target_path)
             newfile.write(content)
             written_files.append(target_path)
 
+
+def build_cookbook(build_config, templatepack_path,
+                   cookbooks_home, force=False):
+    """Build a cookbook from a fastfood.json file.
+
+    Can build on an existing cookbook, otherwise this will
+    create a new cookbook for you based on your templatepack.
+    """
+
+    with open(build_config) as cfg:
+        cfg = json.load(cfg)
+
+    cookbook_name = cfg['name']
+    template_pack = pack.TemplatePack(templatepack_path)
+
+    written_files = []
+    cookbook = create_new_cookbook(cookbook_name, cookbooks_home)
+
+    for stencil_definition in cfg['stencils']:
+
+        selected_stencil_set_name = stencil_definition.get('stencil_set')
+        stencil_set = template_pack.load_stencil_set(selected_stencil_set_name)
+
+        selected_stencil_name = _determine_selected_stencil(
+            stencil_set,
+            stencil_definition
+            )
+
+        stencil = stencil_set.get_stencil(selected_stencil_name,
+                                          **stencil_definition)
+
+        updated_cookbook = process_stencil(
+            cookbook,
+            cookbook_name,  # in case no metadata.rb yet
+            template_pack,
+            force,
+            stencil_set,
+            stencil,
+            written_files
+            )
+
+    return written_files, updated_cookbook
+
+
+def process_stencil(cookbook, cookbook_name, template_pack,
+                    force_argument, stencil_set, stencil, written_files):
+    """Process the stencil requested, writing any missing files as needed.
+
+    The stencil named 'stencilset_name' should
+    be one of templatepack's stencils.
+    """
+
+    # force can be passed on the command line or forced in a stencil's options
+    force = force_argument or stencil['options'].get('force', False)
+
+    if 'files' not in stencil:
+        stencil['files'] = {}
```

Comment:
Same as https://github.com/rackerlabs/fastfood/pull/74/files#r30841494

------------

This might be a little more simple and resilient:

```
stencil['files'] = stencil['files'] or {}
```